mlflow/statsmodels.py

```diff
@@ -473,61 +452,36 @@ def results_to_dict(results):
-
-        :param original_method: the original function object that will be replaced by this function
-        :return: the new fit function, from which we will be doing a call to the original fit
-                 method at some point
-        """
-
-        def fit(self, *args, **kwargs):
+    def wrapper_fit(original, self, *args, **kwargs):
 
-            should_autolog = False
-            if AutologHelpers.should_autolog:
-                AutologHelpers.should_autolog = False
-                should_autolog = True
+        should_autolog = False
+        if AutologHelpers.should_autolog:
+            AutologHelpers.should_autolog = False
+            should_autolog = True
 
-            if not mlflow.active_run():
-                try_mlflow_log(mlflow.start_run)
-                auto_end_run = True
-            else:
-                auto_end_run = False
+        try:
+            if should_autolog:
+                # This may generate warnings due to collisions in already-logged param names
+                log_fn_args_as_params(original, args, kwargs)
 
-            try:
-                # training model
-                model = original_method(self, *args, **kwargs)
+            # training model
+            model = original(self, *args, **kwargs)
 
-                if should_autolog:
-                    # Log the model
+            if should_autolog:
+                # Log the model
+                if get_autologging_config(FLAVOR_NAME, "log_models", True):
                     try_mlflow_log(log_model, model, artifact_path="model")
 
-                    # Log the most common metrics
-                    if isinstance(model, statsmodels.base.wrapper.ResultsWrapper):
-                        metrics_dict = results_to_dict(model)
-                        try_mlflow_log(mlflow.log_metrics, metrics_dict)
-                        # This may generate warnings due to collisions in already-logged param names
-                        log_fn_args_as_params(original_method, args, kwargs)
-
-                return model
-
-            finally:
-                # Clean the shared flag for future calls in case it had been set here ...
-                if should_autolog:
-                    AutologHelpers.should_autolog = True
+                # Log the most common metrics
+                if isinstance(model, statsmodels.base.wrapper.ResultsWrapper):
+                    metrics_dict = results_to_dict(model)
+                    try_mlflow_log(mlflow.log_metrics, metrics_dict)
 
-                # End current run if it had been created here ...
-                if auto_end_run:
-                    try_mlflow_log(mlflow.end_run)
+            return model
 
-        return fit
+        finally:
+            # Clean the shared flag for future calls in case it had been set here ...
+            if should_autolog:
+                AutologHelpers.should_autolog = True
 
-    if not AutologHelpers.patching_accomplished:
```

Comment:
As discussed in the PR description, patching is idempotent. Further, if an user-addressable patching issue occurs during the first call to `mlflow.autolog()`, users should be able to try patching again.

FWIW, because `statsmodels.api` was not imported by default previously, calling `mlflow.statsmodels.autolog()` without importing this module would fail at `patch_class_tree(...)`, but `patching_accomplished` would already be `True`. This prevented users from fixing the import error and trying again.